crmprtd/align.py

```diff
@@ -8,12 +8,222 @@
 
-def align(sesh, obs_tuples):
-    for time, val, var_name, network_name, station_name, lat, lon \
-     in obs_tuples:
-        # Do checks and make station/variable insertions if necessary/possible
-        # Create and yield pycds.Obs object
-        yield Obs()
+
+log = logging.getLogger(__name__)
+ureg = UnitRegistry()
+Q_ = ureg.Quantity
+
+
+def closest_stns_within_threshold(sesh, lon, lat, threshold):
+    # Select all history entries that match this station
+    log.debug("Searching for matching meta_history entries")
+
+    query_txt = """
+        WITH stns_in_thresh AS (
+            SELECT history_id, lat, lon, Geography(ST_Transform(the_geom,4326)) as p_existing,
+                Geography(ST_SetSRID(ST_MakePoint(:x, :y),4326)) as p_new
+            FROM crmp.meta_history
+            WHERE the_geom && ST_Buffer(Geography(ST_SetSRID(ST_MakePoint(:x, :y), 4326)),:thresh)
+        )
+        SELECT history_id, ST_Distance(p_existing,p_new) as dist
+        FROM stns_in_thresh
+        ORDER BY dist
+""" # noqa
+    q = sesh.execute(query_txt, {
+        'x': lon,
+        'y': lat,
+        'thresh': threshold}
+    )
+    valid_hid = set([x[0] for x in q.fetchall()])
+    log.debug("history_ids in threshold", extra={'hid': valid_hid})
+    return valid_hid
+
+
+def convert_unit(val, src_unit, dst_unit):
+    log.debug('Converting units', extra={'src_unit': src_unit,
+                                         'dst_unit': dst_unit})
+    try:
+        val = Q_(val, ureg.parse_expression(src_unit))  # src
+        val = val.to(dst_unit).magnitude  # dest
+    except Exception as e:
+        log.error('Unable to convert units', extra={'src_unit': src_unit,
+                                                    'dst_unit': dst_unit,
+                                                    'exception': e})
+        raise e
+    log.debug('Converted units')
+    return val
+
+
+def unit_db_check(unit_obs, unit_db, val):
+    log.debug('Check if units match')
+    if unit_obs != unit_db:
+        try:
+            val_conv = convert_unit(val, unit_obs, unit_db)
+        except Exception as e:
+            log.error('Unable to convert units',
+                      extra={'src_unit': unit_obs,
+                             'dst_unit': unit_db,
+                             'exception': e})
+            return None
+        return val_conv
+    else:
+        return val
+
+
+def unit_check(sesh, unit_obs, unit_db, val):
+    log.debug('Check if there are units to compare')
+    if not unit_obs and not unit_db:
+        return None
+    else:
+        return unit_db_check(unit_obs, unit_db, val)
+
+
+def match_station_with_active(sesh, obs_tuple, histories):
+    log.debug('Search for active stations')
+    for history in histories:
+        hist = sesh.query(History).filter(
+            and_(History.id == history.id,
+                 History.sdate != None,
+                 History.edate == None))    # noqa
```

Comment:
The behaviour of the query was different when I used ```is not``` and ```is``` rather than ```!=``` and ```==```.  I found an [answer](https://stackoverflow.com/questions/5602918/select-null-values-in-sqlalchemy) as to why this was happening.  However, ```flake8``` will not pass while using ```is not``` and ```is``` and rather than ignore this PEP8 violation altogether, I chose to ignore it for this case.